### PR TITLE
Streaming Sortformer Diarizer [PR 1]: pre-encode input to Conformer

### PR DESCRIPTION
**Pull request by tango4j**
# What does this PR do ?

This PR updates Fast Conformer module to accept pre-encode input if it is specified by the model.

Pre-encode input to Fast-Conformer module is necessary to maintain speaker cache in a form of pre-encode (attention-free) embedding.

**Collection**: [Note which collection this PR will affect]

# Changelog

`nemo/collections/asr/modules/conformer_encoder.py`

# Usage

Follow the examples on any ASR models that are using Fast Conformer module.

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the ASR

# Additional Information

- This PR will be followed by at least 3 PRs related to Streaming Sortformer Diarizer.

### CODE REVIEW

**Comment by ko3n1g** [comment]
hey @tango4j, could have a look at the test failures?

File: nemo/collections/asr/modules/conformer_encoder.py [hunk 1]
```diff
@@ -210,6 +222,7 @@ def input_types(self):
                 "cache_last_channel": NeuralType(('D', 'B', 'T', 'D'), ChannelType(), optional=True),
                 "cache_last_time": NeuralType(('D', 'B', 'D', 'T'), ChannelType(), optional=True),
                 "cache_last_channel_len": NeuralType(tuple('B'), LengthsType(), optional=True),
+                "pre_encode_input": NeuralType(tuple(), LogitsType(), optional=True),
```
Comment (nithinraok) [inline nemo/collections/asr/modules/conformer_encoder.py:225]: its not logits type. Maybe `EncodedRepresentation`?  Also add suggested shape

File: nemo/collections/asr/modules/conformer_encoder.py [hunk 2]
```diff
@@ -543,20 +617,21 @@ def forward_internal(
         else:
             cur_att_context_size = self.att_context_size
 
-        audio_signal = torch.transpose(audio_signal, 1, 2)
+        if not pre_encode_input:
```
Comment (nithinraok) [inline nemo/collections/asr/modules/conformer_encoder.py:620]: By naming of `pre_encode_input` it sounds like you are saying to the forward function to whether pre-encode the input or not. Correct? So if pre-encode_input is False then you shouldn;t pre-encode but I guess you are doing here? 

So just change this to ` if pre_encode_input`?

**Review by nithinraok — CHANGES_REQUESTED** [review]
Add a test case as well.

File: nemo/collections/asr/modules/conformer_encoder.py [hunk 3]
```diff
@@ -517,20 +541,70 @@ def streaming_post_process(self, rets, keep_all_outputs=True):
...
     @typecheck()
     def forward(
-        self, audio_signal, length, cache_last_channel=None, cache_last_time=None, cache_last_channel_len=None
+        self,
+        audio_signal,
+        length,
+        cache_last_channel=None,
+        cache_last_time=None,
+        cache_last_channel_len=None,
+        pre_encode_input=False,
```
Comment (stevehuang52) [inline nemo/collections/asr/modules/conformer_encoder.py:550]: The name `pre_encode_input` is a bit ambiguous, which seems to suggest that it's used to directly inject the pre_encode features by skipping the pre_encode module. Maybe use something like `return_pre_encode`?

File: nemo/collections/asr/modules/conformer_encoder.py (hunk 2 above)
Comment (stevehuang52) [reply nemo/collections/asr/modules/conformer_encoder.py:620]: > By naming of pre_encode_input it sounds like you are saying to the forward function to whether pre-encode the input or not.

I agree, it's a bit confusing.

In the docstring you say:
"`pre_encode_input` is a boolean indicating whether the input audio_signal should be pre-encoded."

which seems to suggest that `pre_encode_input=True` means the audio signal needs to go through `self.pre_encode`, in which case it should be default to `True` instead of `False`.

However, the docstring. is actually doing the reverse:
```
(1) pre_encode_input = False (default):
      audio_signal is a tensor of shape (batch, time, self._feat_in)
(2) pre_encode_input = True:
    audio_signal is a tensor of shape (batch, time, self.d_model)
```

I think we either need to change the docstring and use `if pre_encode_input` here, or change the name to something like `is_pre_encoded`

Comment (tango4j) [reply nemo/collections/asr/modules/conformer_encoder.py:620]: It occurred to me the same way after wrote the docstring. It could be confusing and not clear.
I think if we apply Nithin's opinion also, `bypass_pre_encode` can be more clear?
`is_pre_encoded` is also confusing - because readers don't know "what is pre encoded"?

File: nemo/collections/asr/modules/conformer_encoder.py [hunk 4]
```diff
@@ -11,7 +11,7 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-
+# pylint: disable=E1101,R0902,R0912,R0913,R0914,R0901,C0302,R0915,C0114
```
[the remark this answers is not in the document]
Comment (tango4j) [reply nemo/collections/asr/modules/conformer_encoder.py:14]: removed.

File: nemo/collections/asr/modules/conformer_encoder.py (hunk 1 above)
Comment (tango4j) [reply nemo/collections/asr/modules/conformer_encoder.py:225]: Changed to `BoolType()`.

File: nemo/collections/asr/modules/conformer_encoder.py [hunk 5]
```diff
@@ -149,13 +161,13 @@ class ConformerEncoder(NeuralModule, StreamingEncoder, Exportable, AccessMixin):
             Defaults to False.
         use_pytorch_sdpa (bool): use torch sdpa instead of manual attention.
             Defaults to False.
-        use_pytorch_sdpa_backends (list[str]): list of backend names to use in sdpa. None or empty list means all backends. e.g. ["MATH"]
-            Defaults to None
+        use_pytorch_sdpa_backends (list[str]): list of backend names to use in sdpa.
```
[the remark this answers is not in the document]
Comment (tango4j) [reply nemo/collections/asr/modules/conformer_encoder.py:164]: added this docstring.

File: nemo/collections/asr/modules/conformer_encoder.py [hunk 6]
```diff
@@ -46,7 +46,15 @@
...
-from nemo.core.neural_types import AcousticEncodedRepresentation, ChannelType, LengthsType, NeuralType, SpectrogramType
+from nemo.core.neural_types import (
+    AcousticEncodedRepresentation,
+    BoolType,
+    ChannelType,
+    LengthsType,
+    LogitsType,
+    NeuralType,
+    SpectrogramType,
+)
```
[the remark this answers is not in the document]
Comment (tango4j) [reply nemo/collections/asr/modules/conformer_encoder.py:57]: removed.

**Comment by tango4j** [comment]
> Add a test case as well.

Now that all the comments are addressed. I have added a separate unit-test class in `test_conformer_encoder.py`. It tests all the possible combinations for `bypass_pre_encode`=False or True - including checking the `ValueError` triggering for wrong input format.

**Comment by chtruong814** [comment]
@tango4j hi, I'm reordering the CI queue a bit and will re-add this PR to the queue later.  Trying to prioritize some release items first.

**Review by nithinraok — APPROVED** [review]
LGTM!
